### PR TITLE
MMP delay logic changes

### PR DESCRIPTION
**Pull request by jhammond-intel**
### Description
RFC MMP delay logic changes from @adilger. These are based on the discussion on https://github.com/zfsonlinux/zfs/issues/7045 and https://github.com/zfsonlinux/zfs/pull/7048. (I have not tested these but I plan to.)

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by adilger** [comment]
I would consider these to be more of a "bug fix" than "new feature", but either way I thought it would be useful to get some feedback sooner rather than later...

It needs to add updates to the man/man5/zfs-module-parameters.5 descriptions (I'm working on this already).

I was thinking about the issue of `zfs_multihost_import_intervals` vs. `zfs_multihost_fail_intervals` inconsistency between nodes.  Is there a reason that these values would be module parameters rather than pool parameters?  It would make sense to have them as pool parameters so that they are consistent between any nodes that are trying to import the pool.  There may be a bit of chicken-and-egg problem, but it should still be able to read these parameters from the dataset without actually doing an import?  That is fodder for a separate patch, however.

File: module/zfs/mmp.c [hunk 1]
```diff
@@ -392,6 +391,30 @@ mmp_thread(void *arg)
 		else
 			next_time = start + MSEC2NSEC(MMP_DEFAULT_INTERVAL);
 
+		/*
+		 * Don't allow fail_intervals larger than import_intervals,
```
Comment (ofaaland) [inline module/zfs/mmp.c]: Would this be better implemented in a callback, so you can reject bad values at the time the user tries to set them?  One such example is: 
* 0582e40 Add callback for zfs_multihost_interval

**Comment by ofaaland** [comment]
> Is there a reason that these values would be module parameters rather than pool parameters? It would make sense to have them as pool parameters so that they are consistent between any nodes that are trying to import the pool. There may be a bit of chicken-and-egg problem, but it should still be able to read these parameters from the dataset without actually doing an import?

The two reasons I remember we both arise from the fact that pool properties are stored in an object in the pool.

1. We wanted to make the go/no-go decision very early in the import process, to minimize the chance that some later patch introduces dangerous code before the MMP check is performed.

2. Getting far enough to read an object from the pool results in the node (sometimes) crashing when the pool is changing.  There are several areas of the ZFS kernel module code that assume no change within the pool during the import process.  We found that assumption was made about vdev count, log device count, and IIRC device size.  There might well be more, these just came up early

I agree, it would be great not to depend on uniform settings among the nodes.

**Comment by adilger** [comment]
What about storing the `zfs_multihost_import_interval` parameter in the per-VDEV nvlist?  This is relatively easily parsed (I did that in the util-linux-ng version of libblkid) and could be accessed very early in the import process without any risk of state changing.

**Comment by ofaaland** [comment]
> What about storing the zfs_multihost_import_interval parameter in the per-VDEV nvlist?

One might end up with different values from different VDEVs, for example if some devices get disconnected from host A but remain connected to host B.  But if we take the MAX() of all values found we would still be guaranteed to have a safe value.

It wouldn't be a pool property in a sense, but the user would think of it that way.  Sort of breaks the model, so I'm not sure how to best cope with that.

I agree that storage location for `zfs_multihost_import_interval` has advantages.

**Comment by ofaaland** [comment]
Looking into this I see other issues in MMP that are not dangerous but make MMP more likely to suspend the pool than they should be, or more difficult to debug than it need to be.  I'm working through those and will be back with more information soon.

File: module/zfs/mmp.c [hunk 2]
```diff
@@ -392,6 +391,30 @@ mmp_thread(void *arg)
...
+		if (mmp_fail_intervals >= zfs_multihost_import_intervals) {
+			static boolean_t warned = B_FALSE;
+
+			if (!warned) {
+				cmn_err(CE_WARN,
+				    "zfs_multihost_fail_intervals %u >= "
+				    "zfs_multihost_import_interval %u, "
+				    "using smaller value",
+				    zfs_multihost_fail_intervals,
+				    zfs_multihost_import_interval);
```
Comment (arturpzol) [inline module/zfs/mmp.c]: zfs_multihost_import_interval -> zfs_multihost_import_intervals ?

**Comment by behlendorf** [comment]
I'm going to close out this PR for now.  Several of the changes proposed have already been committed as part of other patches.  The remaining changes in behavior could benefit for additional discussion in a new PR.

**Pull request closed without merging**